quit --waiting command hang up

I think there is a critical issue after upgrade to tornado>5 .
```
circusctl --endpoint=xxx quit --waiting 
```
will successfully stop the process but hang up in terminal. 
After trying to debug this, I find the quit action stuck in https://github.com/circus-tent/circus/blob/2bedab09145a4959ce02b93777dbffbce56edd8e/circus/client.py#L142 lead to timeout. 

To my surprise, `quit` without `--waitting` is all fine.  

This issue can simply reproduced by running `circusd examples/example2.ini` and then stop arbiter with:
```
circusctl --timeout 10000 --endpoint tcp://127.0.0.1:5555 quit --waiting
```

I spent several hours trying to debug this and unfortunately I have no clues why it hangs. It is obvious that there are some events that are not sent and poller is not receiving them, but I don't know which ones...

As for the removal of the `waiting` property - it seems that you are right, there is no point of waiting when using synchronous client.

What I am worried about is that even without `waiting` I get random zmq errors when circusd stops. Those appeared after upgrade to Tornado 5, ex.:

```
2020-07-25 23:27:09 asyncio[7773] [ERROR] Exception in callback BaseAsyncIOLoop._handle_events(14, 1)
handle: <Handle BaseAsyncIOLoop._handle_events(14, 1)>
Traceback (most recent call last):
  File ".../asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File ".../site-packages/tornado/platform/asyncio.py", line 139, in _handle_events
    handler_func(fileobj, events)
  File ".../site-packages/zmq/eventloop/zmqstream.py", line 452, in _handle_events
    zmq_events = self.socket.EVENTS
  File ".../site-packages/zmq/sugar/attrsettr.py", line 48, in __getattr__
    return self._get_attr_opt(upper_key, opt)
  File ".../site-packages/zmq/sugar/attrsettr.py", line 52, in _get_attr_opt
    return self.get(opt)
  File "zmq/backend/cython/socket.pyx", line 472, in zmq.backend.cython.socket.Socket.get
  File "zmq/backend/cython/socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Socket operation on non-socket
2020-07-25 23:27:09 circus[7773] [INFO] Stats streamer stopped
```

--- a/circus/arbiter.py
+++ b/circus/arbiter.py
@@ -571,8 +571,18 @@ class Arbiter(object):
 
     @synchronized("arbiter_stop")
     @gen.coroutine
-    def stop(self):
-        yield self.__stop(True)
+    def stop(self, for_shutdown=True):
+        logger.info('Arbiter exiting')
+        self._stopping = True
+        yield self._stop_watchers(close_output_streams=True,
+                                  for_shutdown=for_shutdown)
+        if self._provided_loop:
+            cb = self.stop_controller_and_close_sockets
+            self.loop.add_callback(cb)
+        else:
+            # stop_controller_and_close_sockets will be
+            # called in the end of start() method
+            self.loop.add_callback(self.loop.stop)
 
     @gen.coroutine
     def _emergency_stop(self):
@@ -760,8 +770,9 @@ class Arbiter(object):
             watchers = self.iter_watchers(reverse=False)
         else:
             watchers = watcher_iter_func(reverse=False)
-        yield [w._stop(close_output_streams, for_shutdown)
-               for w in watchers]
+        futures = [w._stop(close_output_streams, for_shutdown)
+                   for w in watchers]
+        yield gen.multi(futures)
 
     @synchronized("arbiter_stop_watchers")
     @gen.coroutine
@@ -780,8 +791,22 @@ class Arbiter(object):
     @synchronized("arbiter_restart")
     @gen.coroutine
     def restart(self, inside_circusd=False, watcher_iter_func=None):
-        yield self._restart(inside_circusd=inside_circusd,
-                            watcher_iter_func=watcher_iter_func)
+        if inside_circusd:
+            self._restarting = True
+            logger.info('Arbiter exiting')
+            self._stopping = True
+            yield self._stop_watchers(close_output_streams=True,
+                                      for_shutdown=False)
+            if self._provided_loop:
+                cb = self.stop_controller_and_close_sockets
+                self.loop.add_callback(cb)
+            else:
+                # stop_controller_and_close_sockets will be
+                # called in the end of start() method
+                self.loop.add_callback(self.loop.stop)
+        else:
+            yield self._stop_watchers(watcher_iter_func=watcher_iter_func)
+            yield self._start_watchers(watcher_iter_func=watcher_iter_func)
 
     @property
     def endpoint_owner_mode(self):

--- a/circus/watcher.py
+++ b/circus/watcher.py
@@ -782,10 +782,11 @@ class Watcher(object):
         """
         active_processes = self.get_active_processes()
         try:
-            yield [self.kill_process(process,
-                                     stop_signal=stop_signal,
-                                     graceful_timeout=graceful_timeout)
-                   for process in active_processes]
+            futures = [self.kill_process(process,
+                       stop_signal=stop_signal,
+                       graceful_timeout=graceful_timeout)
+                       for process in active_processes]
+            yield gen.multi(futures)
         except OSError as e:
             if e.errno != errno.ESRCH:
                 raise

